d2l-dom-focus.html

```diff
@@ -0,0 +1,160 @@
+<link rel="import" href="d2l-dom.html">
+<link rel="import" href="d2l-dom-visibility.html">
+<script>
+(function() {
+	'use strict';
+
+	var Focus = {
+
+		_focusableElements: {
+			a: true,
+			body: true,
+			button: true,
+			frame: true,
+			iframe: true,
+			img: true,
+			input: true,
+			isindex: true,
+			object: true,
+			select: true,
+			textarea: true
+		},
+
+		getFocusableDescendant: function(node, includeHidden) {
+
+			var composedChildren = D2L.Dom.getComposedChildren(node);
+			//
+
+			for (var i = 0; i < composedChildren.length; i++) {
+
+				if (composedChildren[i].nodeType !== 1) {
+					continue;
+				}
+
+				if (this.isFocusable(composedChildren[i], includeHidden)) {
+					return composedChildren[i];
+				}
+
+				var focusable = this.getFocusableDescendant(composedChildren[i], includeHidden);
+				if (focusable) {
+					return focusable;
+				}
+
+			}
+
+			return null;
+
+		},
+
+		getNextFocusable: function(node, includeHidden) {
+
+			if (!node) {
+				return null;
+			}
+
+			if (includeHidden === undefined) {
+				includeHidden = false;
+			}
+
+			var _getNextAnscestorSibling = function(node) {
+
+				var parentNode = D2L.Dom.getComposedParent(node);
+				while (parentNode) {
+
+					var nextParentSibling = parentNode.nextElementSibling;
+					if (nextParentSibling) {
+						return nextParentSibling;
+					}
+
+					parentNode = D2L.Dom.getComposedParent(parentNode);
+				}
+				return null;
+
+			};
+
+			var _getNextFocusable = function(node, ignore, ignoreChildren) {
+
+				if (!ignore && this.isFocusable(node, includeHidden)) {
+					return node;
+				}
+
+				if (!ignoreChildren) {
+					var focusable = this.getFocusableDescendant(node, includeHidden);
+					if (focusable) {
+						return focusable;
+					}
+				}
+
+				var nextSibling = node.nextElementSibling;
+				if (nextSibling) {
+					var siblingFocusable = _getNextFocusable(nextSibling, false, false);
+					if (siblingFocusable) {
+						return siblingFocusable;
+					}
+
+					return null;
+				}
+
+				var nextParentSibling = _getNextAnscestorSibling(node);
+				if (nextParentSibling) {
+					var parentSibingFocusable = _getNextFocusable(nextParentSibling, false, false);
+					if (parentSibingFocusable) {
+						return parentSibingFocusable;
+					}
+				}
+
+				return null;
+
+			}.bind(this);
+
+			var focusable = _getNextFocusable(node, true, false);
+			return focusable;
+
+		},
+
+		isFocusable: function(node, includeHidden) {
+
+			if (!node || node.disabled) {
+				return false;
+			}
+
+			var _isFocusable = false;
+
+			// IE treats all nodes without tabindex explicitly set as having tabindex=0
+			if (node.getAttributeNode) {
+				var tabIndexAttr = node.getAttributeNode('tabindex');
+				if (tabIndexAttr && tabIndexAttr.specified && tabIndexAttr.value >= 0) {
+					_isFocusable = true;
+				}
+			}
+
+			if (!_isFocusable) {
```

Comment:
I think the is a bug here for case where tabindex is explicitly set to -1 for an element that would normally be focusable.  Tests for the explicit tabindex case are still needed.